SproxydKeysScan/Logging.js

```diff
@@ -0,0 +1,110 @@
+const werelogs = require('werelogs');
+const { env } = require('./env');
+
+/**
+ * @class
+ * @classdesc - aggregates events in a map and logs them at an interval.
+ * call .run() after instantiation to being loop.
+ */
+class AggregateLogger {
+    /**
+     * @constructor
+     * @param {number} logInterval - log summary statistics once every intervalSize seconds
+     */
+    constructor(logInterval) {
+        this.log = new werelogs.Logger('s3utils:SproxydKeysScan:AggregateLogger');
+        if (AggregateLogger._instance) {
+            const message = 'AggregateLogger is a singleton and has been instatiated. This instantiation is ignored';
+            AggregateLogger._instance.log.debug(message);
+            return AggregateLogger._instance;
+        }
+        this.interval = logInterval;
+        this.beginTime = new Date();
+        this.statistics = new Map();
+        AggregateLogger._instance = this;
+        return AggregateLogger._instance;
+    }
+
+    /**
+     * @param {string} event - increases frequency of event string by 1 in this.statistics.
+     * @returns {undefined}
+     */
+    update(event) {
+        this.statistics.set(event, (this.statistics.get(event) || 0) + 1);
+    }
+
+    /**
+     * logs the current event frequencies in this.statistics and resets the counts for the next interval.
+     * @returns {undefined}
+     */
+    logInterval() {
+        const currentTime = new Date();
+        const summary = {};
+        for (const [k, v] of this.statistics) {
+            summary[k] = v;
+        }
+        const data = {
+            beginTime: this.beginTime.toISOString(),
+            endTime: currentTime.toISOString(),
+            summary,
+        };
+        this.log.info('Summary update:', data);
+
+        // setup for next interval
+        this.beginTime = currentTime;
+        this.statistics.clear();
+    }
+
+    run() {
```

Comment:
I feel like calling run in the constructor also works, I can't see a reason to create a "timed logger" and not start the timer right away?

Reply:
I forgot to this to the doc string, but this approach allows usage with both an interval or with event driven logging by calling logInterval explicitly

Reply:
I updated the doc string to explain. If you don't call it right away, this gives you flexibility to either log at an interval or log in response to an event by calling logInterval explicitly.